### PR TITLE
Python Wrapper: Implement Object

### LINKED ISSUE #7007: Python Wrapper: Implement Object Class
image

### Requires refactoring of the ReadableObject and Writeable object classes

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7007

## Change Description

Implement the Object Class

### Background

Refactored WriteObject / ReadObject classes and fixed modeling issues
Testing and bug fixes. 
Implemented missing methods

### Testing Details

Unit and integration testing added

### CODE REVIEW

File: clients/python-wrapper/lakefs/object.py [hunk 1]
```diff
@@ -2,19 +2,380 @@
...
+        """
+        headers = {
+            "Accept": "application/json",
+            "Content-Type": content_type if content_type is not None else "application/octet-stream"
+        }
+
+        # Create user metadata headers
+        if metadata is not None:
+            for k, v in metadata.items():
+                headers[_LAKEFS_METADATA_PREFIX + k] = v
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:320]: validate that the underlaying call_api code validate the header encoding - we like to get an error or encode the right value as part of the header name.
Example posting key='a:key' will probably fail.

File: clients/python-wrapper/lakefs/object.py [hunk 2]
```diff
@@ -2,19 +2,380 @@
...
+        resp = self._client.sdk_client.objects_api.api_client.request(
+            method="PUT",
+            url=url,
+            body=content,
+            headers=headers
+        )
+
+        etag = WriteableObject._extract_etag_from_response(resp.getheaders())
+        staging_metadata = StagingMetadata(staging=staging_location,
+                                           size_bytes=len(content),
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:369]: len of content is always size bytes?

File: clients/python-wrapper/tests/utests/test_object.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                                lambda *args: (_ for _ in ()).throw(lakefs_sdk.exceptions.NotFoundException(
+                                    status=http.HTTPStatus.NOT_FOUND)))
+            assert not obj.exists()
+
+            # Other exception
+            monkeypatch.setattr(lakefs_sdk.api.ObjectsApi, "head_object",
+                                lambda *args: 1 / 0)
+            try:
+                obj.exists()
+                assert 0, "Exception not raised"
```
Comment (nopcoder) [inline clients/python-wrapper/tests/utests/test_object.py:89]: Think that `assert False` will be more readable

**Review by nopcoder — CHANGES_REQUESTED** [review]
Looks good - couple of comments and questions.

File: clients/python-wrapper/lakefs/object.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:369]: Yes? Trying to understand whether this is a trick question... 🤔

File: clients/python-wrapper/lakefs/object.py [hunk 4]
```diff
@@ -2,19 +2,380 @@
...
+            ServerException for any other errors
+        """
+        if read_bytes and read_bytes <= 0:
+            raise OSError("read_bytes must be a positive integer")
+
+        stat = self._obj.stat()
+        if self._pos >= stat.size_bytes:
+            raise EOFError
+        read_bytes = read_bytes if read_bytes is not None else stat.size_bytes
+        new_pos = min(self._pos + read_bytes, stat.size_bytes)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:209]: But you need the new position before you read to provide the range you want to read

File: clients/python-wrapper/lakefs/object.py [hunk 5]
```diff
@@ -2,19 +2,380 @@
...
+            ObjectNotFoundException if repo id, reference id or object path does not exist
+            PermissionException if user is not authorized to perform this operation, or operation is forbidden
+            ServerException for any other errors
+        """
+        if read_bytes and read_bytes <= 0:
+            raise OSError("read_bytes must be a positive integer")
+
+        stat = self._obj.stat()
+        if self._pos >= stat.size_bytes:
+            raise EOFError
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:207]: It should work - but I wanted to simplify the behavior of the wrapper and avoid dealing with the logic of reading negative range and the unexpected behavior

[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:207]: It simplifies the code and avoids potential bugs

File: clients/python-wrapper/lakefs/object.py [hunk 6]
```diff
@@ -2,19 +2,380 @@
...
+        binary_mode = 'b' in mode
+        if binary_mode and isinstance(data, str):
+            content = data.encode('utf-8')
+        elif not binary_mode and isinstance(data, bytes):
+            content = data.decode('utf-8')
+
+        is_presign = pre_sign if pre_sign is not None else self._client.storage_config.pre_sign_support
+        with api_exception_handler(_io_exception_handler):
+            stats = self._upload_presign(content, content_type, metadata) if is_presign \
+                else self._upload_raw(content, content_type, metadata)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:276]: I disagree with this statement here - this is the pythonic way to do it.
IMHO this code statement is very clear

File: clients/python-wrapper/lakefs/object.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:320]: Can you explain why this example will fail?

File: clients/python-wrapper/lakefs/object.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply clients/python-wrapper/lakefs/object.py:209]: You can calculate the new position after your read.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: clients/python-wrapper/lakefs/object.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:321]: Verified - the httplib throws a ValueError exception

File: clients/python-wrapper/lakefs/object.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:370]: Gotcha, fixed and added test

File: clients/python-wrapper/lakefs/object.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py]: Done

File: clients/python-wrapper/lakefs/object.py [hunk 7]
```diff
@@ -2,19 +2,378 @@
...
+            OSError if read_bytes is non-positive
+            ObjectNotFoundException if repo id, reference id or object path does not exist
+            PermissionException if user is not authorized to perform this operation, or operation is forbidden
+            ServerException for any other errors
+        """
+        if read_bytes and read_bytes <= 0:
+            raise OSError("read_bytes must be a positive integer")
+
+        end_pos = self._pos + read_bytes - 1 if read_bytes is not None else ""
+        read_range = _RANGE_STR_TMPL.format(start=self._pos, end=end_pos)
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:206]: 1. end_pos is sometimes a number and sometimes a string - suggest to use None.
2. we always format a read range for the get request. we can make it optional if we also check the start case.

```python
def get_range_string(start, end=None):
    if start == 0 and end is not None:
        return None
    if end is None:
        return f"bytes={start}-"
    return f"bytes={start}-{end}"
```

File: clients/python-wrapper/lakefs/object.py [hunk 8]
```diff
@@ -202,19 +202,15 @@ def read(self, read_bytes: int = None) -> TextIO | BinaryIO:
...
         with api_exception_handler(_io_exception_handler):
             contents = self._client.sdk_client.objects_api.get_object(self._obj.repo,
                                                                       self._obj.ref,
                                                                       self._obj.path,
                                                                       range=read_range,
                                                                       presign=self._pre_sign)
-        self._pos = new_pos  # Update pointer position
+        self._pos += len(contents)  # Update pointer position
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:213]: do we need to add the content len after the decode?

**Review by nopcoder — CHANGES_REQUESTED** [review]
two minor comments about the len and using number/string for end.

File: clients/python-wrapper/lakefs/object.py (hunk 7 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:206]: 1. None formats to the literal 'None'
2. 👍🏽

File: clients/python-wrapper/lakefs/object.py (hunk 8 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:213]: contents returned in bytes so positions should advance by len(contents)

Comment (nopcoder) [reply clients/python-wrapper/lakefs/object.py:219]: Right - my bad - this is return.

**Review by nopcoder — APPROVED** [review]
LGTM!

**Pull request merged**